rust-toolchain.toml

```diff
@@ -0,0 +1,16 @@
+[toolchain]
+# Pinned due to https://github.com/rust-osdev/uefi-rs/issues/500.
+#
+# Compilation started failing 2022-09-02. The previous several builds
+# don't have miri available though, so pin to a slightly older version.
+channel = "nightly-2022-08-26"
+components = [
+    # Needed for `cargo xtask miri`.
+    "miri",
```

Comment:
I tested this locally, manually removing one of the components from the minimal profile (e.g. `rustup component remove cargo-x86_64-unknown-linux-gnu --toolchain nightly-2022-08-26-x86_64-unknown-linux-gnu`) does indeed cause the build to break. However, I'm not sure it's worth guarding against that case since uninstalling cargo or rustc is not easy to do accidentally. Also the error message to correct the problem is pretty good:

```
$ cargo xtask build
error: 'cargo' is not installed for the toolchain 'nightly-2022-08-26-x86_64-unknown-linux-gnu'
To install, run `rustup component add cargo --toolchain nightly-2022-08-26-x86_64-unknown-linux-gnu`
```

One thing that was confusing me is why I needed to specify the clippy and rustfmt components at all, since they are part of the default profile. Turns out that the [github runners are set to the minimal profile by default](https://github.com/actions/runner-images/issues/600), which makes sense. To keep things more minimal for CI, I've removed all the components here except for `rust-src`, and in the github CI I added some explicit `rustup component add` steps for the few jobs that need extra components.

I think this should hit the right balance of "everything works locally by default" and "CI downloads only what it needs to".

Reply:
What do we win by not listing all default components + miri? Time? Like, 1-2s per build? I think convenience here is more important than convenience.

>  I'm not sure it's worth guarding against that case since uninstalling cargo or rustc is not easy to do accidentally.

Yeah, but what about the case where someone was on stable all the time, builds uefi-rs the first time, but doesn't get a nightly version of cargo, clippy, and rustc?

I'm in favor of adding all default components + miri

Reply:
> Yeah, but what about the case where someone was on stable all the time, builds uefi-rs the first time, but doesn't get a nightly version of cargo, clippy, and rustc?

Ah, I wasn't fully clear before with what I tested. I also tested the case you describe here, and it behaves correctly: all the default profile stuff gets installed. You only get a build error if you manually remove a required component. So the only thing that doesn't get installed by default is miri, which I think is fine as we don't actually test _that_ much stuff with miri, and I don't think most people are normally running the local miri tests. Also, miri is often missing from recent builds (see https://rust-lang.github.io/rustup-components-history/), so we get better "recent-nightly" coverage if we don't require that component when not necessary.

Reply:
So I'm confused now. Right now, it's only 

```toml
components = [
    # Needed for `-Zbuild-std`.
    "rust-src",
]
```

Does this match to your description above? I'm not sure if those components are handled as additional components regarding the default profile or not. I always list all of them explicitely to be safe.. any comment?

Reply:
Yep, that's the intended set of components. The components list is indeed additive (I think I should probably put up a PR in the rustup docs to make that explicit). Here's are some notes on expected behavior, let me know if you spot any cases that don't seem like we're doing the right thing:

* Local development:
  * User has a default setup (rustup is installed, and just the stable channel is installed):
    * Running any `cargo` command in the repo will install the toolchain specified in our `rust-toolchain.toml`. In a normal rustup setup the default profile is `default`, so all of those components get installed (`cargo clippy rust-docs rust-std rustc rustfmt`). The `rust-toolchain.toml` file also specifies `rust-src`, so that gets installed automatically too.
    * `miri` isn't installed by default, which is fine because it's not that commonly run. If they do try to run `cargo xtask miri`, they'll get a helpful message from `rustup` saying what command to run to add that component.
  * User has their profile set to `minimal`:
    * Pretty much the same as above, but they won't automatically get some useful tools like `clippy`. I think we should respect the user's configuration here though, presumably they chose to set the `minimal` profile for a reason. If they try to run a command like `cargo xtask clippy`, they automatically get a helpful error message from `rustup` with the command to add the `clippy` component to the toolchain.
  * User has the desired toolchain installed, but they have manually removed some important component like `rustc`:
    * Same deal as above, they will get a helpful error message saying how to add the component.
* CI:
  * The default profile is `minimal`. That is sufficient for most of the CI jobs that just need `cargo` and `rustc`. For jobs that require additional components (`miri`, `rustfmt`, and `clippy`) we explicitly add that component with call to `rustup`.

Reply:
I found it experimentally. I've attempted to look through the rustup code for a more definitive answer, but didn't find the right place so far.

I did put up a PR in rustdoc to expand the docs for rust-toolchain: https://github.com/rust-lang/rustup/pull/3067